### PR TITLE
FE_Q1_NonLocal

### PR DESCRIPTION
**Pull request by davydden**
this does not work yet (thus Draft PR).

### CODE REVIEW

**Comment by davydden** [comment]
@luca-heltai please let me know when the other PR supports distribution of DoFs, I will rebase/squash this one and see if I can make it work.

**Comment by luca-heltai** [comment]
@davydden , my branch is compiling now. I had to simplify the interface to the fe, beacuse I was not able to get the get_non_local_dofs_on_cell(cell) to work properly with any of the TriaAccessor, TriaAccessorBase, CellAcessor, etc. 

None of them was compiling, so at some point I just put an index there.

I'll try again later on (I was interrupted too many times).

Distribute dofs should now do the correct thing.

**Comment by luca-heltai** [comment]
(that is, provided that you know how to go from the cell index to index set of the non local dofs)

**Comment by davydden** [comment]
good, I will rebase and see if I can do something here...

**Comment by davydden** [comment]
@luca-heltai 

> (that is, provided that you know how to go from the cell index to index set of the non local dofs)

eventually we should make cell iterators work, there is no easy way to go from cell index to cell iterator.

**Comment by davydden** [comment]
@luca-heltai `DoFHadler::distribute()` runs, but `n_dofs()` returns `0`, I guess we forgot to adjust it.

**Comment by davydden** [comment]
@luca-heltai So `distribute()` compiles and runs, but it doe snot distribute correctly ;-(( As cell DoFs we get `numbers::invalid_unsigned_int`. Nonetheless, I think we can merge this into you branch to continue there (I think only you can change the PR status form `draft` to `read for review`).

File: tests/fe/fe_q1_nonlocal_01.cc [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      std::vector<types::global_dof_index> vertex_indices(
+        GeometryInfo<2>::vertices_per_cell);
+      for (auto cell = dof_handler.begin_active(); cell != dof_handler.end();
+           ++cell)
+        {
+          cell->get_dof_indices(local_dof_indices);
+          for (unsigned int v = 0; v < GeometryInfo<2>::vertices_per_cell; ++v)
+            {
+              vertex_indices[v] = cell->vertex_index(v);
+              // AssertDimension(local_dof_indices[v], vertex_indices[v]);
```
Comment (davydden) [inline tests/fe/fe_q1_nonlocal_01.cc]: that's the failing part

**Comment by davydden** [comment]
closing this one for now.

**Pull request closed without merging**